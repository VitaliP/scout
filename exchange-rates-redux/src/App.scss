.App {

}

.main {
  margin: 100px auto;
  max-width: 800px;
}

.navbar {
  /*background-color: #f0f1f2;*/
}

.loading-lbl {
  margin: 0 0 0 10px;
}

.modal-container {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 1;
  background-color: rgba(0,0,0,0.5);

  .modal-dialog {
    margin-top: 150px;
  }

}


