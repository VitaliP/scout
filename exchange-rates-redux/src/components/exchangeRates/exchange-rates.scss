button:focus {outline:0;}

.controls, .title {
  margin: 0 0 15px 0;
}

.title {
  font-size: 20px;
  line-height: 26px;
}

select {
  margin: 0 20px 0 0;
  height: 40px;
}

.react-datepicker-wrapper {
  margin-right: 20px;
  input {
    width: 100px;
    height: 40px;
    border: 1px solid #aaa;
    border-radius: 5px;
    background-color: #f8f8f8;
    padding-left: 8px;
  }
}

.btn, select, .react-datepicker-wrapper {
  vertical-align: top;
}

.btn {
  margin: 1px 0 0 20px;
}

.lbl {
  line-height: 40px;
  margin-right: 5px;
  display: inline-block;
}

.controls {
  padding: 20px 0 20px;
  border-top: 1px solid #aaa;
  text-align: right;

  .btn {
    margin: 0 0 0 0;
  }
}

.selector {
  margin-bottom: 20px;
}