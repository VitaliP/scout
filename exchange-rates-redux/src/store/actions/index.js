export const FETCH_DATA = 'FETCH_DATA';
export const SCREEN_STATE = 'SCREEN_STATE';
// export const FETCH_ERROR = 'FETCH_ERROR';

export const MODAL_SHOW = 'MODAL_SHOW';
export const MODAL_HIDE = 'MODAL_HIDE';