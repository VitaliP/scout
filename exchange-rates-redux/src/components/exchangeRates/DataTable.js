import _ from 'lodash';
import React from 'react'
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import './exchange-rates.scss';
import { bindActionCreators } from 'redux'


const prepareData = (rates, sortBy, sortDirection) => {

    let arrRates = _.chain(rates)
        .toPairs()
        .map((item)=>{
            return {
                currency: item[0],
                rate: item[1]
            };
        })
        .sortBy(sortBy)
        .value();

    if (sortDirection === 'desc') {
        arrRates.reverse();
    }

    return arrRates;

};


const DataTable = (props) => {

    const rates = prepareData(props.rates, props.sort.sortBy, props.sort.sortDirection);

    if (!rates || rates.length<1) {
        console.error("no data? why?");
        return false;
    }

    return (

        <table className="table">
            <thead>
            <tr>
                <th scope="col">Currency</th>
                <th scope="col">Rate</th>
            </tr>
            </thead>
            <tbody>

            {rates.map((row,index)=>{
                return (
                    <tr key={index}>
                        <td>{row.currency}</td>
                        <td>{row.rate}</td>
                    </tr>
                )

            })}

            </tbody>
        </table>

    );

};

DataTable.propTypes = {
    rates: PropTypes.object.isRequired,
    sort: PropTypes.object.isRequired
};

const mapState = state => {

    return {
        sort: state.exchangeRates.sort,
        rates: state.exchangeRates.rates
    }
};

const mapDispatchToProps = dispatch => bindActionCreators(
    {},
    dispatch,
);

export default connect(mapState, mapDispatchToProps)(DataTable)