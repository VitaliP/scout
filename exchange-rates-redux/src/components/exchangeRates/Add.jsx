import React, { Component } from 'react'
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { newExchangeRate } from '../../store/actions/exchangeRates.actions'
import { bindActionCreators } from 'redux'

class AddForm extends Component {

    constructor() {
        super();

        this.onSubmit = this.onSubmit.bind(this);
    }

    onSubmit(e) {
        e.preventDefault();

        const currency = this.refs['currency'].value;
        const rate = this.refs['rate'].value;


        this.props.newExchangeRate(currency, rate);

        return false;

    }

    render() {


        return (
            <div className="add">

                <form onSubmit={this.onSubmit}>

                    <div>
                        <label htmlFor="currency">
                            Currency:<br />
                            <input name="currency" ref="currency" />
                        </label>
                    </div>

                    <div>
                        <label htmlFor="rate">
                            Rate:
                            <br />
                            <input name="rate" ref="rate" />
                        </label>
                    </div>

                    <div>
                        <button type="submit" className="btn btn-primary">Submit</button>
                    </div>



                </form>




            </div>
        )
    }


}

AddForm.propTypes = {
    newExchangeRate: PropTypes.func.isRequired
};

const mapState = state => {
    return {}
};


const mapDispatchToProps = dispatch => bindActionCreators(
    {
        newExchangeRate
    },
    dispatch,
);


export default connect(mapState, mapDispatchToProps)(AddForm)