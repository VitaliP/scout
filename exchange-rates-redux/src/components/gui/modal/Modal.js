import React, { Component } from 'react'
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { modalHide } from '../../../store/actions/modal.actions';

class Modal extends Component {

    constructor(props) {
        super(props);

        this.hideModal = this.hideModal.bind(this);
        this.preventPropogation = this.preventPropogation.bind(this);
    }

    onFetchClick() {
        this.props.fetchData();
    }

    hideModal() {
        this.props.modalHide();
    }

    preventPropogation(e) {
        e.stopPropagation();
    }

    render() {
        if (!this.props.modalState) return false;

        let modalTitle;
        switch (this.props.params.type) {
            case 'alert': modalTitle = "Alert"; break;
            default: modalTitle = 'unknown'; break;
        }

        return <div className="modal-container" onClick={this.hideModal} data-test="modal">

            <div className="modal-dialog" role="document" onClick={this.preventPropogation}>
                <div className="modal-content">
                    <div className="modal-header">
                        <h5 className="modal-title" data-test="title">{modalTitle}</h5>
                        <button type="button" className="close" data-dismiss="modal" aria-label="Close" onClick={this.hideModal}>
                            <span aria-hidden="true">&times;</span>
                        </button>
                    </div>
                    <div className="modal-body">
                        <p data-test="msg">{this.props.params.message}</p>
                    </div>
                    <div className="modal-footer">
                        <button type="button" className="btn btn-primary" onClick={this.hideModal}>OK</button>
                    </div>
                </div>
            </div>

        </div>;

    }
}

Modal.propTypes = {
    modalState: PropTypes.bool.isRequired,
    params: PropTypes.object
};

const mapStateToProps = state => {
    return {
        modalState: state.modal.modalState,
        params: state.modal.params
    }

};


export default connect(mapStateToProps, {modalHide})(Modal)