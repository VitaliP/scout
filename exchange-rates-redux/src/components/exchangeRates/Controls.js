import React from 'react'
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { sortAction } from '../../store/actions/exchangeRates.actions'
import { bindActionCreators } from 'redux'

const Controls = (props) => {

    const field = props.sort.sortBy;
    const direction = props.sort.sortDirection;

    const sortByCurrency = () => props.sortAction('currency', direction);
    const sortByRate = () => props.sortAction('rate', direction);

    const sortAsc = () => props.sortAction(field, 'asc');
    const sortDesc = () => props.sortAction(field, 'desc');

    let classesSortByCurrency = 'btn btn-sm ';
    let classesSortByRate = 'btn btn-sm ';
    let classesSortAsc = 'btn btn-sm ';
    let classesSortDesc = 'btn btn-sm ';
    classesSortByCurrency += (field === 'currency')?'btn-secondary':'btn-link';
    classesSortByRate += (field === 'rate')?'btn-secondary':'btn-link';
    classesSortAsc += (direction === 'asc')?'btn-secondary':'btn-link';
    classesSortDesc += (direction === 'desc')?'btn-secondary':'btn-link';

    return (
        <div className="controls">
            <button type="button" className={classesSortByCurrency} onClick={sortByCurrency}>Sort by Currency</button>
            &nbsp;
            <button type="button" className={classesSortByRate} onClick={sortByRate}>Sort by Rate</button>
            &nbsp;
            &nbsp;
            &nbsp;
            <button type="button" className={classesSortAsc} onClick={sortAsc}>ASC</button>
            &nbsp;
            <button type="button" className={classesSortDesc} onClick={sortDesc}>DESC</button>
        </div>
    )
};

Controls.propTypes = {
    sortAction: PropTypes.func.isRequired,
    sort: PropTypes.object.isRequired
};

const mapState = state => {

    return {
        sort: state.exchangeRates.sort
    }

};


const mapDispatchToProps = dispatch => bindActionCreators(
    {
        sortAction
    },
    dispatch,
);


export default connect(mapState, mapDispatchToProps)(Controls)