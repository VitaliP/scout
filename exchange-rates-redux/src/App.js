import React from 'react';

import { Provider } from 'react-redux';
import store from './store/store';

import ErrorBoundary from './components/common/ErrorBoundary';
import Navbar from './components/gui/navigation/Navbar';
import ExchangeRates from './components/exchangeRates/ExchangeRates';
import Modal from './components/gui/modal/Modal'

function App() {
    return (

        <Provider store={store}>

            <div className="App" data-test="app">

                <ErrorBoundary>
                    <Navbar />
                </ErrorBoundary>

                <ErrorBoundary>
                    <ExchangeRates />
                </ErrorBoundary>



            </div>

            <Modal />

        </Provider>



    );
}

export default App;
