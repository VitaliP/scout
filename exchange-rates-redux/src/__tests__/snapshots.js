import React from 'react';
import { Provider } from 'react-redux';
import store from '../store/store';
import App from '../App';
import Modal from '../components/gui/modal/Modal';
import { mount } from 'enzyme';
import { fetchData } from '../store/actions/exchangeRates.actions'
import moxios from 'moxios';

describe("UI snapshots",()=>{

    let wrapper = mount(<Provider store={store}><App /><Modal /></Provider>);

    beforeEach(()=>{
        moxios.install();
    });

    afterEach(()=>{
        moxios.uninstall();
    });


    it('Initial state', () => {
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Table with real data', () => {
        moxios.wait(() => {
            const request = moxios.requests.mostRecent();
            request.respondWith({
                status: 200,
                response: {"base":"EUR","rates":{"BGN":1.9558,"NZD":1.7259,"ILS":4.0549,"RUB":72.1432,"CAD":1.5069,"USD":1.1234,"PHP":58.703,"CHF":1.1214,"ZAR":16.6463,"AUD":1.6355,"JPY":122.06,"TRY":6.595,"HKD":8.7963,"MYR":4.6913,"THB":35.196,"HRK":7.4055,"NOK":9.7938,"IDR":16103.94,"DKK":7.4674,"CZK":25.56,"HUF":322.29,"GBP":0.89208,"MXN":21.5292,"KRW":1333.25,"ISK":141.5,"SGD":1.54,"BRL":4.3796,"PLN":4.2605,"INR":78.577,"RON":4.728,"CNY":7.7794,"SEK":10.6495},"date":"2019-06-17"}
            })
        });

        return store.dispatch(fetchData())
            .then(() => {
                expect(wrapper.html()).toMatchSnapshot();
            });
    });

});