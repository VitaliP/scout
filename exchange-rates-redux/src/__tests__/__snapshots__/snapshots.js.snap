// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI snapshots Initial state 1`] = `"<div class=\\"App\\" data-test=\\"app\\"><nav class=\\"navbar navbar-expand-lg navbar-light bg-light\\" data-test=\\"nav-bar\\"><div class=\\"navbar-brand\\" data-test=\\"app-name\\">ER.redux</div><button class=\\"navbar-toggler\\" type=\\"button\\" data-toggle=\\"collapse\\" data-target=\\"#navbarSupportedContent\\" aria-controls=\\"navbarSupportedContent\\" aria-expanded=\\"false\\" aria-label=\\"Toggle navigation\\"><span class=\\"navbar-toggler-icon\\"></span></button><div class=\\"collapse navbar-collapse\\" id=\\"navbarSupportedContent\\"><ul class=\\"navbar-nav mr-auto\\"><li class=\\"nav-item active\\"><a class=\\"nav-link\\" href=\\"/\\">Home <span class=\\"sr-only\\">(current)</span></a></li></ul></div></nav><div class=\\"main\\"><button type=\\"button\\" class=\\"btn btn-primary\\" data-test=\\"btn-fetch\\">Fetch data</button></div></div>"`;

exports[`UI snapshots Table with real data 1`] = `"<div class=\\"App\\" data-test=\\"app\\"><nav class=\\"navbar navbar-expand-lg navbar-light bg-light\\" data-test=\\"nav-bar\\"><div class=\\"navbar-brand\\" data-test=\\"app-name\\">ER.redux</div><button class=\\"navbar-toggler\\" type=\\"button\\" data-toggle=\\"collapse\\" data-target=\\"#navbarSupportedContent\\" aria-controls=\\"navbarSupportedContent\\" aria-expanded=\\"false\\" aria-label=\\"Toggle navigation\\"><span class=\\"navbar-toggler-icon\\"></span></button><div class=\\"collapse navbar-collapse\\" id=\\"navbarSupportedContent\\"><ul class=\\"navbar-nav mr-auto\\"><li class=\\"nav-item active\\"><a class=\\"nav-link\\" href=\\"/\\">Home <span class=\\"sr-only\\">(current)</span></a></li></ul></div></nav><div class=\\"main\\"><table class=\\"table\\"><thead><tr><th scope=\\"col\\">EUR x</th><th scope=\\"col\\">Currency</th><th scope=\\"col\\">Rate</th></tr></thead><tbody><tr><td></td><td>AUD</td><td>1.6355</td></tr><tr><td></td><td>BGN</td><td>1.9558</td></tr><tr><td></td><td>BRL</td><td>4.3796</td></tr><tr><td></td><td>CAD</td><td>1.5069</td></tr><tr><td></td><td>CHF</td><td>1.1214</td></tr><tr><td></td><td>CNY</td><td>7.7794</td></tr><tr><td></td><td>CZK</td><td>25.56</td></tr><tr><td></td><td>DKK</td><td>7.4674</td></tr><tr><td></td><td>GBP</td><td>0.89208</td></tr><tr><td></td><td>HKD</td><td>8.7963</td></tr><tr><td></td><td>HRK</td><td>7.4055</td></tr><tr><td></td><td>HUF</td><td>322.29</td></tr><tr><td></td><td>IDR</td><td>16103.94</td></tr><tr><td></td><td>ILS</td><td>4.0549</td></tr><tr><td></td><td>INR</td><td>78.577</td></tr><tr><td></td><td>ISK</td><td>141.5</td></tr><tr><td></td><td>JPY</td><td>122.06</td></tr><tr><td></td><td>KRW</td><td>1333.25</td></tr><tr><td></td><td>MXN</td><td>21.5292</td></tr><tr><td></td><td>MYR</td><td>4.6913</td></tr><tr><td></td><td>NOK</td><td>9.7938</td></tr><tr><td></td><td>NZD</td><td>1.7259</td></tr><tr><td></td><td>PHP</td><td>58.703</td></tr><tr><td></td><td>PLN</td><td>4.2605</td></tr><tr><td></td><td>RON</td><td>4.728</td></tr><tr><td></td><td>RUB</td><td>72.1432</td></tr><tr><td></td><td>SEK</td><td>10.6495</td></tr><tr><td></td><td>SGD</td><td>1.54</td></tr><tr><td></td><td>THB</td><td>35.196</td></tr><tr><td></td><td>TRY</td><td>6.595</td></tr><tr><td></td><td>USD</td><td>1.1234</td></tr><tr><td></td><td>ZAR</td><td>16.6463</td></tr></tbody></table></div></div>"`;
